### PR TITLE
FEAT: Added field warning component to all fields

### PR DESCRIPTION
**Pull request by obulaworld**
#### Summary

Update all Field component to handle warning texts.
It seeks to fully fix this UI-KIT [issue](https://github.com/commercetools/ui-kit/issues/2671)

## Description

This PR seeks to update all Field component to handle warning texts by:
- Adding the new `FieldWarnings` component to all Fields
- Updating all fields dependencies to include FieldWarnings package
- Updating all fields story, spec, visual route to reflect the new changes.

## Screenshots
Screenshot 2024-01-23 at 10 56 28
Screenshot 2024-01-23 at 10 56 38
Screenshot 2024-01-23 at 10 56 50

### CODE REVIEW

**Comment by obulaworld** [comment]
I had to replace all select fields' hasWarning prop to only be true if and only if the field is touched and the newly added warnings object prop has at least one warning. Please let me know if this will affect any team that might have been using the hasWarning prop as a simple truthy or falsy prop.

File: .changeset/fair-cooks-press.md
Comment (CarlosCortizasCT) [inline .changeset/fair-cooks-press.md:23]: Perhaps we can elaborate a little more this message so consumers get a better idea about this change:

```suggestion
We've included a new property in all `*Field` components so consumers can not only render errors (below the input) but also warnings.
The main difference between them is the color used for the text.

Please take a look at the RADME files of the fields components to learn how to use this new property.
```

File: packages/components/fields/async-creatable-select-field/README.md
Comment (CarlosCortizasCT) [inline packages/components/fields/async-creatable-select-field/README.md:63]: I'm not sure this is a property we want to expose in the `Field` components.
I know we have this property in the `FieldWarnings` component (similar to what we have in the `FieldErrors` one), but I don't see we are using it for errors so I think we shouldn't neither for warnings.

@emmenko do you maybe remember why is this prop used for (in the `FieldErrors` component?
I don't find any component in this repository using the `FieldErrors` component and passing it the `renderDefaultWarning` prop.

Comment (emmenko) [reply packages/components/fields/async-creatable-select-field/README.md:63]: I have no idea what the `renderDefaultError` is meant for.

Maybe we should just deprecate it and remove it, the `renderError` should be good enough.

Comment (CarlosCortizasCT) [reply packages/components/fields/async-creatable-select-field/README.md:63]: Thanks @emmenko 

@obulaworld so in this PR we can at least remove the `renderDefaultWarning` from the `FieldWarnings` component (as we were not using it anywhere yet) and don't use it in the fields components.

Does it make sense to you?

Comment (obulaworld) [reply packages/components/fields/async-creatable-select-field/README.md:63]: Yes @CarlosCortizasCT It does. I will remove it.

**Comment by CarlosCortizasCT** [comment]
It seems we're missing something in the `DateRageField` component as it's not rendering the orange border when it has a warning message.

image

**Comment by obulaworld** [comment]
> It seems we're missing something in the `DateRageField` component as it's not rendering the orange border when it has a warning message.
> 
> image

@CarlosCortizasCT Fixed

File: packages/components/fields/date-field/src/date-field.visualroute.jsx
Comment (CarlosCortizasCT) [inline packages/components/fields/date-field/src/date-field.visualroute.jsx:92]:
```suggestion
    <Spec label="with warning when touched">
```

File: packages/components/fields/date-time-field/src/date-time-field.visualroute.jsx
Comment (CarlosCortizasCT) [inline packages/components/fields/date-time-field/src/date-time-field.visualroute.jsx:101]:
```suggestion
    <Spec label="with warning when touched">
```

File: packages/components/fields/localized-multiline-text-field/src/localized-multiline-text-field.visualroute.jsx
Comment (CarlosCortizasCT) [inline packages/components/fields/localized-multiline-text-field/src/localized-multiline-text-field.visualroute.jsx:141]:
```suggestion
    <Spec label="with warning when touched">
```

File: packages/components/fields/localized-text-field/src/localized-text-field.visualroute.jsx
Comment (CarlosCortizasCT) [inline packages/components/fields/localized-text-field/src/localized-text-field.visualroute.jsx:117]:
```suggestion
    <Spec label="with warning when touched">
```

File: packages/components/fields/money-field/src/money-field.visualroute.jsx
Comment (CarlosCortizasCT) [inline packages/components/fields/money-field/src/money-field.visualroute.jsx:131]:
```suggestion
    <Spec label="with warning when touched">
```

File: packages/components/fields/multiline-text-field/src/multiline-text-field.visualroute.jsx
Comment (CarlosCortizasCT) [inline packages/components/fields/multiline-text-field/src/multiline-text-field.visualroute.jsx:85]:
```suggestion
    <Spec label="with warning when touched">
```

**Review by CarlosCortizasCT — COMMENTED** [review]
I think we just to adjust some test names and we will be good to go

**Review by CarlosCortizasCT — APPROVED** [review]
Many thanks for your work in this PR, Chisom 🙇 

Please, wait for Filip and another Shield engineer approval 🙏

File: packages/components/fields/async-creatable-select-field/src/async-creatable-select-field.tsx
Comment (kark) [inline packages/components/fields/async-creatable-select-field/src/async-creatable-select-field.tsx:492]: I'm a bit confused about one aspect in general. Based on my understanding, the original requirements were outlined [here](https://github.com/commercetools/ui-kit/issues/2671#issuecomment-1886715030).

I read [this comment](https://github.com/commercetools/ui-kit/pull/2698#issuecomment-1905713921) but I'm not sure if this was addressed. 

Previously, the `hasWarning` prop alone triggered the border display, but it's no longer considered. Is the `hasWarning` prop deprecated? Shouldn't this be mentioned in the changeset?

**Review by kark — COMMENTED** [review]
Thank you, I'm leaving the comment to get a better understanding of the changes

File: packages/components/fields/async-creatable-select-field/src/async-creatable-select-field.tsx
Comment (CarlosCortizasCT) [reply packages/components/fields/async-creatable-select-field/src/async-creatable-select-field.tsx:492]: That's a good catch, Kacper.

I think the `hasWarning` prop we currently have should be deprecated, but that means keeping it so consumers have time to transition from one prop (`hasWarning`) to the other (`warnings`).

We probably need the `hasWarning` constant calculation to be adjusted:
```js
const hasWarning =
     this.props.hasWarnings ||
      (AsyncCreatableSelectInput.isTouched(this.props.touched) &&
      hasWarnings(this.props.warnings));
```

And also update the jsdoc of that property to mention about it.

@kark what do you think?

Comment (obulaworld) [reply packages/components/fields/async-creatable-select-field/src/async-creatable-select-field.tsx:492]: Ahhh yeahh I think it should. @CarlosCortizasCT Whats your take on this?

Comment (obulaworld) [reply packages/components/fields/async-creatable-select-field/src/async-creatable-select-field.tsx:492]: Cool stuff. @CarlosCortizasCT So the jsdoc should mention that the field is already deprecated or is their a specific copy you would like to use?

Comment (CarlosCortizasCT) [reply packages/components/fields/async-creatable-select-field/src/async-creatable-select-field.tsx:493]: Hi @obulaworld 

> @CarlosCortizasCT So the jsdoc should mention that the field is already deprecated or is their a specific copy you would like to use?

I mean both.
For example:
```
  /**
   * Indicates the input field has a warning.
   * @deprecated Please use the `warnings` prop instead so users know the reason why the field is in warning state.
   */
  hasWarning?: boolean;
```

> Also does this mean we keep the hasWarning boolean prop in the story?

We should remove it as we don't want consumers to use it anymore.

Comment (obulaworld) [reply packages/components/fields/async-creatable-select-field/src/async-creatable-select-field.tsx:494]: Done

**Review by CarlosCortizasCT — APPROVED** [review]
Thanks.

File: packages/components/fields/async-creatable-select-field/src/async-creatable-select-field.visualroute.jsx
Comment (kark) [inline packages/components/fields/async-creatable-select-field/src/async-creatable-select-field.visualroute.jsx:88]: @CarlosCortizasCT Do you think we should delete the `<Spec>`s with the legacy `hasWarning` or leave them?

**Review by kark — COMMENTED** [review]
Thank you for the contribution 🎉 looks good to me

File: packages/components/fields/async-creatable-select-field/src/async-creatable-select-field.visualroute.jsx
Comment (obulaworld) [reply packages/components/fields/async-creatable-select-field/src/async-creatable-select-field.visualroute.jsx:88]: @CarlosCortizasCT ?

Comment (obulaworld) [reply packages/components/fields/async-creatable-select-field/src/async-creatable-select-field.visualroute.jsx:88]: @kark I think we can remove them when we are finally removing everything about the hasWarning prop.

Comment (CarlosCortizasCT) [reply packages/components/fields/async-creatable-select-field/src/async-creatable-select-field.visualroute.jsx:88]: Yes, I think we should keep them to make sure we don't break that behaviour ahead of time.
We still want to support that prop although we want to encourage consumers to not use it anymore.

Comment (obulaworld) [reply packages/components/fields/async-creatable-select-field/src/async-creatable-select-field.visualroute.jsx:88]: Okay. @CarlosCortizasCT I do not have the permission to either approve the percy test or merge this PR. So I believe someone on the UI-Kit will have to do both

Comment (CarlosCortizasCT) [reply packages/components/fields/async-creatable-select-field/src/async-creatable-select-field.visualroute.jsx:88]: > Many thanks for your work in this PR, Chisom 🙇
> 
> Please, wait for Filip and another Shield engineer approval 🙏

As I explained before, we need another Shield engineer approval as well as @FilPob's one.

**Comment by FilPob** [comment]
Looking good from my side! I noticed only one point: In the `Timefield` the orange border takes precedence, even though both errors and warnings are shown. In all other fields when both errors and warnings are shown, the Red border takes precedence, which is correct
image

File: packages/components/fields/async-creatable-select-field/src/async-creatable-select-field.visualroute.jsx
Comment (kark) [reply packages/components/fields/async-creatable-select-field/src/async-creatable-select-field.visualroute.jsx:88]: > @kark I think we can remove them when we are finally removing everything about the hasWarning prop.

My point here was that we are removing a bunch of `<Specs>` in this PR e.g.:
Screenshot 2024-01-29 at 16 43 04
and we should probably leave them for now

Comment (obulaworld) [reply packages/components/fields/async-creatable-select-field/src/async-creatable-select-field.visualroute.jsx:88]: Ahhhhh Okay. I get it now. Will add them back plus the new ones

**Comment by obulaworld** [comment]
> Looking good from my side! I noticed only one point: In the `Timefield` the orange border takes precedence, even though both errors and warnings are shown. In all other fields when both errors and warnings are shown, the Red border takes precedence, which is correct image

This will be coming from the Input directly. So I guess I will update the Input in this PR.

**Review by kark — APPROVED** [review]

**Pull request merged**
